server/src/main/scala/org/http4s/server/middleware/authentication/NonceKeeperF.scala

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2014 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s.server.middleware.authentication
+
+import cats.effect.Clock
+import cats.effect.Concurrent
+import cats.effect.Timer
+import cats.effect.concurrent.Ref
+import cats.effect.concurrent.Semaphore
+import cats.syntax.all._
+
+import java.util.LinkedHashMap
+import java.{util => ju}
+import scala.concurrent.duration.MILLISECONDS
+
+private[authentication] object NonceKeeperF {
+  def apply[F[_]: Timer](
+      staleTimeout: Long,
+      nonceCleanupInterval: Long,
+      bits: Int,
+  )(implicit F: Concurrent[F]): F[NonceKeeperF[F]] = for {
+    // This semaphore controls who has access to `nonces` during stale nonce eviction. This must never be set above one.
+    semaphore <- Semaphore[F](1)
+    currentTime <- Clock[F].realTime(MILLISECONDS)
```

Comment:
`realTime` is prone to non-monotonicity.  If this is about intervals and not about wall time, we might want the monotonic time.